build.sbt

```diff
@@ -0,0 +1,31 @@
+name := "sttp-openapi-generator-example"
+
+version := "0.1"
+
+scalaVersion := "2.13.2"
+
+val root = file(".")
+
+lazy val petstoreApi = project
+  .in(file("petstore-api"))
+  .settings(
+    openApiInputSpec := s"${baseDirectory.value.getPath}/petstore.yaml",
+    openApiGeneratorName := "scala-sttp",
+    openApiOutputDir := baseDirectory.value.name,
+    openApiIgnoreFileOverride := s"$root/openapi-ignore-file",
+    libraryDependencies ++= Seq(
+      "com.softwaremill.sttp.client" %% "core" % "2.2.0",
+      "com.softwaremill.sttp.client" %% "json4s" % "2.2.0",
+      "org.json4s" %% "json4s-jackson" % "3.6.8"
+    ),
+    (compile in Compile) := ((compile in Compile) dependsOn openApiGenerate).value,
+    cleanFiles += baseDirectory.value / "src"
```

Comment:
I think we should generate those into src-managed. See: https://www.scala-sbt.org/1.x/docs/Howto-Generating-Files.html

Maybe we can also leverage the `sourceGenerators` task?

Reply:
That looks awesome, unfortunately we can't use `sourceGenerators` because this requires all files to be compile-able and despite explicitly saying to generator that we want only `src/main` files it also generates `.openapi-generator-ignore` and `.openapi-generator/VERSION` files ...

I also agree that we should use `sourceManaged`, but while it works I am not sure if it is safe since the generator still doesn't allow us change the destination of the source files - we can only change the baseDir directory.
So it looks like this:
```bash
$ tree petstore-api/target/scala-2.12  
petstore-api/target/scala-2.12
├── src_managed
│   └── main
│       └── demo
│           └── src
│               └── main
│                   └── scala
│                       └── org
│                           └── openapitools
│                               └── client
│                                   ├── api
│                                   │   ├── EnumsSerializers.scala
│                                   │   ├── PetApi.scala
│                                   │   ├── StoreApi.scala
│                                   │   └── UserApi.scala
│                                   ├── core
│                                   │   ├── ApiInvoker.scala
│                                   │   ├── requests.scala
│                                   │   └── Serializers.scala
│                                   └── model
│                                       ├── ApiResponse.scala
│                                       ├── Category.scala
│                                       ├── Order.scala
│                                       ├── Pet.scala
│                                       ├── Tag.scala
│                                       └── User.scala
```


So after all it seems like this ignoreFile isn't what we needed.

Reply:
I wonder what implications would be of doing it in the openapi-way. We could generate a self-containing sbt-project into petstore-api directory and since it contains build.sbt file it will be recognized by the root project, won't it?

Reply:
> I wonder what implications would be of doing it in the openapi-way. We could generate a self-containing sbt-project into petstore-api directory and since it contains build.sbt file it will be recognized by the root project. am I right?

yes, but it would also contain its own settings, which would overwrite user's settings. We don't want to do that - the plugin should only do its job: generate a client, and minimally interfere with other parts of the project

Reply:
> That looks awesome, unfortunately we can't use `sourceGenerators` because this requires all files to be compile-able and despite explicitly saying to generator that we want only `src/main` files it also generates `.openapi-generator-ignore` and `.openapi-generator/VERSION` files ...
>
> I also agree that we should use `sourceManaged`, but while it works I am not sure if it is safe since the generator still doesn't allow us change the destination of the source files - we can only change the baseDir directory.

Ah, so that's a bug in the openapi sbt-plugin or the generator. We'll have to live with that so far :). So maybe we can leave this as-is, but add a comment on why `sourceGenerators` are not used, and link to issues (if they exist) which report the problem?

> So after all it seems like this ignoreFile isn't what we were looking for.

Well it is, but needs fixing, no? ;)

Reply:
> Ah, so that's a bug in the openapi sbt-plugin or the generator. We'll have to live with that so far :). So maybe we can leave this as-is, but add a comment on why sourceGenerators are not used, and link to issues (if they exist) which report the problem?

Agree, currently there is no issue apart from https://github.com/OpenAPITools/openapi-generator/issues/6685 but I think that it will evolve into new one probably in sbt-openapi-plugin.

> Well it is, but needs fixing, no? ;)
I think that it is something which by coincidence was partially useful for us, but I cannot image it being changed in a such way that will allow us to put sources into srcManaged directly. I wasn't just build for that purpose.

Reply:
I just filed a PR against OpenAPITools that would allow use of source generators by disabling the metadata files in a way compatible with the sbt plugin.

If that's interesting to you still it could use some support!

https://github.com/OpenAPITools/openapi-generator/pull/7862